src/codegen/events/substrate.rs

```diff
@@ -32,68 +47,131 @@ impl EventEmitter for SubstrateEventEmitter<'_> {
-                            cfg,
-                            contract_no,
-                            Some(func),
-                            self.ns,
-                            vartab,
-                            opt,
-                        );
-
-                        topics.push(e);
-                        topic_tys.push(Type::Bytes(32));
-                    }
-                    _ => {
-                        let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                        topics.push(e);
-                        topic_tys.push(ty);
-                    }
-                }
-            } else {
-                let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                data.push(e);
-                data_tys.push(arg.ty());
+        let loc = pt::Loc::Builtin;
+        let event = &self.ns.events[self.event_no];
+        // For freestanding events the name of the emitting contract is used
+        let contract_name = &self.ns.contracts[event.contract.unwrap_or(contract_no)].name;
+        let hash_len = || Box::new(Expression::NumberLiteral(loc, Type::Uint(32), 32.into()));
+
+        // Events that are not anonymous always have themselves as a topic.
+        // This is static and can be calculated at compile time.
+        if !event.anonymous {
+            let mut encoded = format!("{}::{}", contract_name, &event.name).encode();
+            encoded[0] = 0; // Set the prefix (there is no prefix for the event topic)
+            topics.push(Expression::AllocDynamicBytes(
+                loc,
+                Type::Slice(Type::Uint(8).into()),
+                hash_len(),
+                Some(topic_hash(&encoded[..])),
+            ));
+            topic_tys.push(Type::DynamicBytes);
+        };
+
+        // Topic prefixes are static and can be calculated at compile time.
+        let mut topic_prefixes: VecDeque<Vec<u8>> = event
+            .fields
+            .iter()
+            .filter(|field| field.indexed)
+            .map(|field| {
+                format!(
+                    "{}::{}::{}",
+                    contract_name,
+                    &event.name,
+                    &field.name_as_str()
+                )
+                .into_bytes()
+                .encode()
+            })
+            .collect();
+
+        for (ast_exp, field) in self.args.iter().zip(event.fields.iter()) {
+            let value_exp = expression(ast_exp, cfg, contract_no, Some(func), self.ns, vartab, opt);
+            let value_var = vartab.temp_anonymous(&value_exp.ty());
+            let value = Expression::Variable(loc, value_exp.ty(), value_var);
+            cfg.add(
+                vartab,
+                Instr::Set {
+                    loc,
+                    res: value_var,
+                    expr: value_exp,
+                },
+            );
+            data_tys.push(value.ty());
+            data.push(value.clone());
```

Comment:
You can preallocate `data` and `data_tys` on lines 45 and 46, using `Vec::with_capacity`, for better performance.

Reply:
There is a limit of 4 topics on Substrate. When a rust `Vec` needs to grow, there is a minimum size which is 4 for elements > 1 byte: https://github.com/rust-lang/rust/blob/b7463e8bdb7b039e75241dcc4ea65bbd24d5c608/library/alloc/src/raw_vec.rs#L111-L117